src/cgmemmgr.cpp

```diff
@@ -0,0 +1,839 @@
+        allocations.clear();
+    }
+    // Allocations that have not been finalized yet.
+    SmallVector<Allocation, 16> allocations;
+    void *alloc(size_t size, size_t align)
+    {
+        size_t min_size = (size_t)-1;
+        int min_id = 0;
+        for (int i = 0;i < nblocks && blocks[i].ptr;i++) {
+            auto &block = blocks[i];
+            void *ptr = block.alloc(size, align);
+            if (ptr) {
+                void *wr_ptr;
+                if (block.state & ROBlock::InitAlloc) {
+                    wr_ptr = ptr;
+                }
+                else {
+                    wr_ptr = get_wr_ptr(block, ptr, size, align);
+                }
+                block.state |= ROBlock::Alloc;
+                allocations.push_back(Allocation{wr_ptr, ptr, size, false});
+                return wr_ptr;
+            }
+            if (block.avail < min_size) {
+                min_size = block.avail;
+                min_id = i;
+            }
+        }
+        size_t block_size = get_block_size(size);
+        auto &block = blocks[min_id];
+        auto new_block = alloc_block(block_size);
+        block.swap(new_block);
+        if (new_block.state) {
+            pendings.push_back(std::move(new_block));
+        }
+        else {
+            new_block.reset(nullptr, 0);
+        }
+        void *ptr = block.alloc(size, align);
+#ifdef _OS_WINDOWS_
+        block.state = ROBlock::Alloc;
+        void *wr_ptr = get_wr_ptr(block, ptr, size, align);
+        allocations.push_back(Allocation{wr_ptr, ptr, size, false});
+        ptr = wr_ptr;
+#else
+        block.state = ROBlock::Alloc | ROBlock::InitAlloc;
+        allocations.push_back(Allocation{ptr, ptr, size, false});
+#endif
+        return ptr;
+    }
+};
+
+template<bool exec>
+class DualMapAllocator : public ROAllocator<exec> {
+protected:
+    void *get_wr_ptr(ROBlock &block, void *rt_ptr, size_t, size_t) override
+    {
+        assert((char*)rt_ptr >= block.ptr &&
+               (char*)rt_ptr < (block.ptr + block.total));
+        if (!(block.state & ROBlock::WRInit)) {
+            block.wr_ptr = (uintptr_t)create_shared_map(block.total,
+                                                        block.wr_ptr);
+            block.state |= ROBlock::WRInit;
+        }
+        if (!(block.state & ROBlock::WRReady)) {
+            protect_page((void*)block.wr_ptr, block.total, Prot::RW);
+            block.state |= ROBlock::WRReady;
+        }
+        return (char*)rt_ptr + (block.wr_ptr - uintptr_t(block.ptr));
+    }
+    ROBlock alloc_block(size_t size) override
+    {
+        ROBlock new_block;
+        // use `wr_ptr` to record offset initially
+        auto ptr = alloc_shared_page(size, (size_t*)&new_block.wr_ptr, exec);
+        new_block.reset(ptr, size);
+        return new_block;
+    }
+    void finalize_block(ROBlock &block, bool reset)
+    {
+        if (!(block.state & ROBlock::Alloc)) {
+            if ((block.state & ROBlock::WRInit) && reset)
```

Comment:
I'm not sure what llvm guarantees. This will happen if we are called with a large block that happens to have less free space then other blocks. (e.g. all other blocks has 1k free space and LLVM asked for a 1023.5k block followed by a 2k block).